src/features/customerList/__tests__/CustomerListContainer.test.tsx

```diff
@@ -0,0 +1,124 @@
+import React from 'react';
+
+import { render, screen, userEvent } from '../../../common/testUtils/testingLibrary';
+import CustomerListContainer from '../CustomerListContainer';
+
+type SetupConfig = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  mocks: any[];
+};
+
+function setup(config?: SetupConfig) {
+  const mocks = config?.mocks ?? [];
+
+  render(<CustomerListContainer />, {
+    mocks,
+  });
+
+  return {
+    toggleFilters: () => {
+      userEvent.click(screen.getByRole('button', { name: 'Rajaa' }));
+    },
+  };
+}
+
+test(`customer list filter form has expected fields`, () => {
+  const { toggleFilters } = setup();
+
+  toggleFilters();
+
+  // For performance reasons, cover all these fields in the same test case
+  // without re-initializing the DOM.
+  [
+    {
+      name: 'customerGroups',
+      label: 'Asiakasryhmä',
+      type: 'multiselect',
+    },
+    {
+      name: 'boatTypeIds',
+      label: 'Veneen tyyppi',
+      type: 'multiselect',
+    },
+    {
+      name: 'leaseStatuses',
+      label: 'Sopimuksen tila',
+      type: 'multiselect',
+    },
+    {
+      name: 'berthSection',
+      label: 'Venepaikat',
+      type: 'checkbox',
+    },
+    {
+      name: 'harborIds',
+      label: 'Venesatama',
+      type: 'multiselect',
+    },
+    {
+      name: 'harborIds',
+      label: 'Laituri ja paikka',
+      type: 'select',
+    },
+    {
+      name: 'winterStorageSection',
+      label: 'Talvisäilytysruutupaikat',
+      type: 'checkbox',
+    },
+    {
+      name: 'winterStorageGridAreaIds',
+      label: 'Talvisäilytysruutualue',
+      type: 'multiselect',
+    },
+    {
+      name: 'winterStoragePlaceId',
+      label: 'Talvisäilytyspaikka',
+      type: 'select',
+    },
+    {
+      name: 'unmarkedWinterStorageSection',
+      label: 'Nostojärjestysalue',
+      type: 'checkbox',
+    },
+    {
+      name: 'winterStorageAreaIds',
+      label: 'Nostojärjestysalue',
+      type: 'multiselect',
+    },
+    {
+      name: 'dateSection',
+      label: 'Aikaväli:',
+      type: 'checkbox',
+    },
+    // No good way to find these
+    // {
+    //   name: 'dateStart',
+    //   label: '',
+    //   type: 'date',
+    // },
+    // {
+    //   name: 'dateEnd',
+    //   label: '',
+    //   type: 'date',
+    // },
+    {
+      name: 'moreThanOneBerthOrWinterStorage',
+      label: 'Enemmän kuin yksi vene- tai talvisäilytyspaikka',
+      type: 'checkbox',
+    },
+  ].forEach(({ label, type }) => {
+    switch (type) {
+      case 'multiselect':
+        expect(screen.getByRole('combobox', { name: label })).toBeInTheDocument();
+        break;
+      case 'select':
+        expect(screen.getByRole('combobox', { name: label })).toBeInTheDocument();
+        break;
+      case 'checkbox':
+        expect(screen.getByRole('checkbox', { name: label })).toBeInTheDocument();
```

Comment:
`screen.queryByRole` is more suitable for `.toBeInDocument` assertion. But also `getByRole` will tell you which combobox wasn't found. In a way (code execution) `.toBeDocument` is redundant when using `getByRole`.

In this situation `getByRole` might be better.

Reply:
I think I know what you mean. I had this same thought for some time, but I've maybe changed my mind since.

The expectation will be truthy 99% of the time the codes executes that far because most of the times that something is wrong, the getter will throw.

As I see it, the expectation (and `toBeInTheDocument`) is there more for the human. It helps communicate that we are making an assertion on this line. Without `expect` the line could seem bare to some eyes. The error `getBy` yields is helpful in pointing the dev towards the correct direction when they have to adjust tests.

Checking out the readme for `jest-dom`, their first example of this helper uses a `getBy` selector:
https://github.com/testing-library/jest-dom#tobeinthedocument

Also, testing library itself has this approach documented:
https://testing-library.com/docs/guide-disappearance/#2-using-waitfor

It gives `queryBy` as an option when asserting that the element isn't in the document (which is the most common usecase for `queryBy`):
https://testing-library.com/docs/guide-disappearance/#nottobeinthedocument

I look at the `getBy` selectors' throwing as a way to make fixing tests easier for the developer. That's why I don't mind using them in this way anymore.

Reply:
I completely agree and was thinking the same when commenting (which is why my comment was unnecessary :D). `getByRole` for debuggability and `.toBeInTheDocument` for readability.